### ISSUE TITLE
Question regarding combining `output_norm.weight`  and `blk.0.attn_norm.weight` at the point of joining?

### ISSUE THREAD

**Issue by jukofyork** [arcee-ai/mergekit#147]
Just wondered if you could help point me in the right direction with this:

I'm experimenting with merging coding models and it's become clear that the only real method that works is to stack repeated copies of the whole models. Trying to do Goliath style interleave merging very quickly shows the "spelling mistake" problem and I hypothesize this is due to different feature abstraction level(s) being moved about (eg: it's like a convolution network's early layers that look like Gabor filters getting merged with 'cat detectors' deeper in the network...).

So testing the stacked coding models it's clear they are actually better, as can be seen by giving huge sections of code and asking to refactor - often the existing coding models will either copy the code verbatim and/or say they have made changes that they haven't (using fp16 vs q8_0 seems to stop some of this).

I also think the stacked models  seem to have greater "understanding" when asked to explain code, for example:

**deepseek-coder:33b-instruct-fp16**
```
This Java code is a part of an Eclipse plugin named "Assistai". The class ChatConversationManager is responsible for managing the chat conversation between the user and the AI assistant. Here's a breakdown of its objectives:

1. Initialize: It initializes the chat conversation, sets up listeners for log messages and preference changes, and creates an instance of OllamaModelManager.

2. Undo and Clear Operations: It provides methods to undo the last user message and clear all messages in the conversation.

3. Stop Operation: It cancels all running jobs related to the chat conversation.

4. Send User Messages: It sends a user message to the assistant, updates the view with the new message, and schedules a job to send the conversation to the assistant.

5. Update Assistant Messages: It updates the assistant's messages in the view.

6. Find Message View: It finds the ChatConversationView in the application model by its element ID.

7. Send Predefined Prompts: It sends a predefined prompt to the assistant based on the active editor and its context.

8. Display Notification and Error Messages: It displays notification and error messages in the chat view.

Design choices:

1. Singleton: The class is annotated with @Singleton, indicating that there should be only one instance of this class throughout the application. This is a common design pattern used to ensure that only one instance of a class is created and provides a global point of access to it.

2. Dependency Injection: The class uses dependency injection (DI) to inject its dependencies, which makes the code more modular, testable, and easier to maintain.

3. Logging and Error Handling: It sets up listeners for log messages and handles errors by displaying them in the chat view. This is a common practice in many applications to provide feedback to the user about what's happening behind the scenes.

4. Job Scheduling: It uses the Eclipse job scheduling framework to schedule jobs for sending the conversation to the assistant. This allows the application to continue running while the job is being executed, improving responsiveness.

5. View Finding: It provides a static method findMessageView() to find the ChatConversationView in the application model by its element ID. This is a common pattern in Eclipse plugins for finding views based on their IDs.

6. Context Extraction: It extracts context information from the active editor, such as the filename, language, code snippet, errors, clipboard text, and user input. This allows the assistant to provide more relevant responses.

7. Predefined Prompts: It sends predefined prompts based on the type of prompt requested. This is a common feature in chatbots where the system can send pre-defined messages based on certain triggers or commands.

8. Language Tag Extraction: It extracts a language tag from the file extension to provide more accurate syntax highlighting and code execution.

Improvements:

1. Error Handling: The error handling could be improved by using custom exceptions instead of generic ones. This would make it easier to catch and handle specific types of errors in other parts of the application.

2. Testability: The class could be made more testable by extracting some of its methods into separate classes or interfaces. This would allow for easier unit testing.

3. Concurrency: The class uses a synchronized block to ensure that only one job can run at a time. However, this could lead to blocking if the job takes a long time to complete. A better solution would be to use a thread pool executor with a bounded queue size to limit the number of concurrent jobs.

4. Context Extraction: The context extraction could be improved by extracting more information from the active editor, such as the cursor position or the selected text. This would allow the assistant to provide more relevant responses.

5. Predefined Prompts: The predefined prompts could be made more flexible and customizable. For example, they could accept parameters that customize the prompt based on the user's input.
```

**deepseek-coder:2x33b-instruct-q8_0**
```
This is a Java class, ChatConversationManager, which is part of an Eclipse plugin for an AI-assisted coding environment. The main purpose of this class is to manage the chat conversation between the user and the AI assistant. 

Here's a high-level rundown of what the class does:

1. Initialization: It's annotated with @Creatable and @Singleton, which means it's a "Creatable" Eclipse component and is a Singleton, meaning there's only one instance of this class in the application. 

2. Dependency Injection: The class has fields that are annotated with @Inject. This is a form of Dependency Injection (DI), where the framework (Eclipse) automatically provides the necessary dependencies when it's time to create an instance of this class.

3. Post-Construction Initialization: The @PostConstruct method is used to set up the class after it's been fully constructed. This includes setting up a log listener, which will print any warnings or errors to the Eclipse log, and a property change listener, which will print out a message when the selected model name in the preferences changes.

4. Undoing, Clearing, and Stopping: The class has methods for undoing, clearing, and stopping the current conversation. 

5. Sending Messages: It can send user-initiated messages, and it can also start, update, and end a message from the AI.

6. Finding the ChatConversationView: The class has a static method for finding the ChatConversationView in the Eclipse workbench. This is used to update the chat conversation in the Eclipse part when new messages are sent or received.

7. Sending Predefined Prompts: It can also send predefined prompts, which are pre-defined messages that the user can send. 

8. Error and Notification Messages: The class has methods for displaying error and notification messages.

9. Utility Methods: The class also includes a method to get the appropriate Markdown language tag for a file, and a method to remove indentation from a string of code. 

This class is a key part of the Eclipse plugin's functionality, as it manages the back-and-forth communication between the user and the AI. It's designed to be easily testable, with clear, well-documented methods for each main function.
```

This seems to be inline with Goliath's  "show not tell" story writing ability!

I can also see this increased "understanding" in the refactored code they produce.

------

I hypothesize that the reason this works might actually just be to do with the first model creating a much richer embedding for the second model to use. (ie: a single model is only seeing 1-hot encoded inputs and can't see the interesting correlations between the tokens, etc).

It seems important to use the same model; possibly due to the trained prompt format, but it could also be that the below problem is hindering my attempts to 'sandwich' the base model in the middle, etc.

The problem I'm seeing is this:

- It should in theory be possible to stack these models endlessly (and if I can get this to work I will actually see if I can hack the `llamap.cpp` code to not have to load identical copies of the same weights!), but this isn't the case and `deepseek-coder:3x33b-instruct-q8_0` starts to get confused...
- Some models like `codelamma` seem a lot more sensitive to this and will start to do strange things, like this:

```
The code you've provided is a Java class that manages a conversation between a human and an AI-powered chatbot, which I will refer to as the "conversational AI (AI-bot for brevity. The main objective of this 

The main objective of this class...
```

So thinking about this  the easiest solution would be to keep the `output_norm.weight` and `token_embd.weight` at the point of joining, but I have a feeling that `llama.cpp` will get confused by this and it probably won't work.

So thinking some more the next thing to try is to combine the `output_norm.weight`  and `blk.0.attn_norm.weight` at the point of joining. (I think) these are vectors of  size `n_embed` containing standard-deviations:

https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html

So statistically they should be combined via `sqrt(x_i^2 + y_i^2)`, but since they are just getting used to divide one after another it might actually be just be `x_i * y_i` (I'll figure this out when I make the changes...).

This isn't quite the same as keeping the final softmax with the `output_norm.weight` and the `token_embd.weight`  weights: this will act as a change of location in log space, but with all these normalization layers and the fact that this is almost working, makes me think this isn't actually very important and the values are probably almost mean-centered anyway...

I'm not asking for you to write any code or make any changes, but would appreciate it if you give me any pointers on where i need to look to hack this change in?

I'll report back if I have any luck and if this does work I will then look at the `llama.cpp` code and see if I can make this use a single copy of the weights.

**Comment by jukofyork** [comment]
I've just thought it might also be important to factor in the norms of the `token_embd.weight` for each of the `n_embed` vectors of size `n_vocab` too (unless there is some reason these might have been standardized already?).

**Comment by jukofyork** [comment]
Actually the more I think about it the less this seems like a good idea as it's essentially trying to squeeze the "transformer space" back to the "embedding space" just to feed back to the first layer, and if it's almost working then there is no reason to believe this will be any more successful. I wonder if it might just be the case that the first layer(s) (and possibly the last layer(s)) have adapted to a different feature representation.

**Comment by jukofyork** [comment]
Do you have any idea what `output_norm.weight` is doing?

From the original "Attention Is All You Need" paper:

> 3.4 Embeddings and Softmax
Similarly to other sequence transduction models, we use learned embeddings to convert the input
tokens and output tokens to vectors of dimension dmodel. We also use the usual learned linear transfor-
mation and softmax function to convert the decoder output to predicted next-token probabilities. *In
our model, we share the same weight matrix between the two embedding layers* and the pre-softmax
linear transformation, similar to [ 30 ]. In the embedding layers, we multiply those weights by √dmodel.

So the output of the final tensor in the final layer `blk.{last}.ffn_norm.weight`  will be vector of size `n_embed`  and from the above it sounds like the transpose of  `token_embd.weight` with dimension (`n_embed`, `n_vocab`) is used or perhaps it could be that things have changed since then and ` output.weight` with dimension (`n_embed`, `n_vocab`) is doing this job now, but...

I can't really see where the `output_norm.weight` vector of size `n_embed` sits in all this:

- It obviously can't just sit at the end as `blk.{last}.ffn_norm.weight` will be sat right before it and will have already standardized it's output...
- It doesn't seem to match up with either the transpose of  `token_embd.weight`  nor the ` output.weight` unless it does indeed sit pointlessly after `blk.{last}.ffn_norm.weight`...

**Comment by jukofyork** [comment]
I forgot to say that removing the **last** transformer block does indeed effect the weird `codellama` output:

```
slices:
  - sources:
    - model: codellama-34b-istruct
      layer_range: [0, 47]
  - sources:
    - model: codellama-34b-istruct
      layer_range: [0, 48]
merge_method: passthrough
base_model: codellama-34b-istruct
tokenizer_source: base
```

```
The provided source code is for a Java class called Ch\atConversationManager that appears to be part of an Eclipse rich client platform (RCP) application. The class is annotated with the @Creatable and @Singleton annotations, which suggest that it is intended to be a reusable and sharable component of the larger application.

Based on the Javadoc comments and the method signatures, it appears that this class is responsible for managing conversations between a human user and an AI assistant, possibly in the context of a chat interface. The class exposes methods for sending and receiving messages, undoing and clearing conversations, and canceling running jobs.

Here are some of the design choices and objectives that can be gleaned from a cursory review of the code:

1. Dependency Injection: The class is designed to be loosely coupled and modular, with dependencies injected through its constructor. This suggests that the author values testability, reusability, and modularity in software design.
2. Single Responsibility Principle: The class appears to adhere to the single responsibility principle, as it is focused on managing conversations and does not appear to have unrelated responsibilities.
3. Cohesion and Coupling: The class has high cohesion, as its methods are tightly related to the core responsibilities of managing conversations. However, the coupling between this class and its collaborators (e.g., ChatMessageFactory, IJobManager) is not immediately apparent from this excerpt.
4. Error Handling: The class does not appear to handle errors or exceptions very gracefully, with some error messages simply being logged and others causing the application to crash. This could be a potential area for improvement in terms of user experience and robustness.
5. Code Readability: Some of the method and variable names could be more descript\
```

vs 

```
The code you've provided is a Java class that manages a conversation between a human and an AI-powered chatbot, which I will refer to as the "conversational AI (AI-bot for brevity. The main objective of this 

The main objective of this class is
```

But obviously it isn't perfect as the `descript\` end shows...

Removing transformer blocks from the beginning like so:

```
slices:
  - sources:
    - model: codellama-34b-istruct
      layer_range: [0, 48]
  - sources:
    - model: codellama-34b-istruct
      layer_range: [1, 48]
merge_method: passthrough
base_model: codellama-34b-istruct
tokenizer_source: base
```
didn't really have much effect and hence why I'm back to looking at the final post-transformer layers again...
